### PR TITLE
Add DOFA model

### PR DESCRIPTION
**Pull request by ShadowXZT**
Adding support for the DOFA, a unified foundation model for Remote sensing data

### CODE REVIEW

**Comment by ShadowXZT** [comment]
@microsoft-github-policy-service agree company="TUM"

microsoft-github-policy-service[bot] ***@***.***>
于2024年2月23日周五 14:58写道：

**Comment by ShadowXZT** [comment]
@microsoft-github-policy-service agree company="TUM"

File: torchgeo/models/dofa.py [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+            d_model=input_dim,
+            nhead=num_heads,
+            activation="gelu",
+            norm_first=False,
+            batch_first=False,
+            dropout=False,
+        )
+        self.transformer_encoder = nn.TransformerEncoder(
+            encoder_layer, num_layers=num_layers, enable_nested_tensor=False
+        )  # type: ignore[no-untyped-call]
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:87]: Missing type hints, added in https://github.com/pytorch/pytorch/pull/120550

File: torchgeo/models/dofa.py [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+    Returns:
+        A DOFA small 16 model.
+    """
+    model = OFAViT(
+        patch_size=16,
+        embed_dim=384,
+        depth=12,
+        num_heads=6,
+        mlp_ratio=4,
+        norm_layer=partial(nn.LayerNorm, eps=1e-6),  # type: ignore[arg-type]
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:446]: mypy doesn't yet support partials: https://github.com/python/mypy/issues/1484

File: torchgeo/models/dofa.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+        """
+        y = self.w1(x)
+        y = self.nonlin1(y)
+        y = self.w2(y)
+        y = self.nonlin2(y)
+        out: Tensor = x + y
+        return out
+
+
+class Dynamic_MLP_OFA(nn.Module):
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:148]: I would like to rename this without the underscores. Maybe `DOFAMLP`?

File: torchgeo/models/dofa.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+        self,
+        wv_planes: int,
+        inter_dim: int = 128,
+        kernel_size: int = 3,
+        embed_dim: int = 1024,
+    ) -> None:
+        """Initialize a new Dynamic_MLP_OFA instance.
+
+        Args:
+            wv_planes: Dimensions of dynamic weight generator.
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:161]: No idea what this is

File: torchgeo/models/dofa.py [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+        self.weight_generator = TransformerWeightGenerator(
+            wv_planes, self._num_kernel, embed_dim
+        )
+        self.scaler = 0.01
+
+        self.fclayer = FCResLayer(wv_planes)
+
+        self._init_weights()
+
+    def _get_weights(self, waves: Tensor) -> Tensor:
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:184]: Probably going to get rid of this helper function, it doesn't do much and is only 1 line of code

File: torchgeo/models/dofa.py [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+            img_feat, weights, bias=bias, stride=self.kernel_size, padding=1, dilation=1
+        )
+
+        x = dynamic_out
+        x = x.flatten(2).transpose(1, 2)
+
+        return x, waves
+
+
+class OFAViT(nn.Module):
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:246]: Should we rename this to DOFA?

File: torchgeo/models/dofa.py [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+        self.head = (
+            nn.Linear(embed_dim, num_classes) if num_classes > 0 else nn.Identity()
+        )
+
+    def forward_features(self, x: Tensor, wave_list: list[float]) -> Tensor:
+        """Forward pass of the feature embedding layer.
+
+        Args:
+            x: Input mini-batch.
+            wave_list: List of wavelengths.
```
Comment (adamjstewart) [inline torchgeo/models/dofa.py:335]: Can we rename this to `wavelengths`? I think it's more clear. The type hints already make it clear that it's a list.

File: torchgeo/models/dofa.py (hunk 6 above)
Comment (ShadowXZT) [reply torchgeo/models/dofa.py:246]: yes, DOFA sounds better

File: torchgeo/models/dofa.py (hunk 7 above)
Comment (ShadowXZT) [reply torchgeo/models/dofa.py:335]: ok

File: tests/models/test_dofa.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        "wavelengths",
+        [
+            # Gaofen
+            [0.443, 0.565, 0.763, 0.765, 0.910],
+            # NAIP
+            [0.640, 0.560, 0.480],
+            [0.480, 0.560, 0.640, 0.810],
+            # Sentinel-1
+            [5.405],
+            [5.405, 5.405],
```
[the remark this answers is not in the document]
Comment (ShadowXZT) [reply tests/models/test_dofa.py:42]: Should we add a comment here: The central wavelength should be in \mu m unit?

[the remark this answers is not in the document]
Comment (adamjstewart) [reply tests/models/test_dofa.py:42]: Technically it only has be to in μm to use the pre-trained weights, you could use any units with the model itself, correct?

File: torchgeo/models/dofa.py (hunk 3 above)
Comment (adamjstewart) [reply torchgeo/models/dofa.py:148]: Which sounds more accurate, DynamicMLPEncoder or DynamicMLPEmbedding?

File: torchgeo/models/dofa.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/models/dofa.py:161]: So it's the number of bands for the satellite?

File: tests/models/test_dofa.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply tests/models/test_dofa.py:42]: Exact numbers don't matter for unit tests. I don't really like that we randomly chose to use 3.75, I think we should change that. I added units to the function docstring.

**Comment by adamjstewart** [comment]
https://github.com/pytorch/pytorch_sphinx_theme/pull/192 will fix the table display. If it doesn't get merged, I'll add the hack locally

**Pull request merged**